ingestion/src/metadata/ingestion/source/pipeline/databrickspipeline/metadata.py

```diff
@@ -0,0 +1,222 @@
+#  Copyright 2021 Collate
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  http://www.apache.org/licenses/LICENSE-2.0
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""
+Databricks pipeline source to extract metadata
+"""
+
+import traceback
+from typing import Any, Iterable, List, Optional
+
+from metadata.generated.schema.api.data.createPipeline import CreatePipelineRequest
+from metadata.generated.schema.api.lineage.addLineage import AddLineageRequest
+from metadata.generated.schema.entity.data.pipeline import (
+    PipelineStatus,
+    StatusType,
+    Task,
+    TaskStatus,
+)
+from metadata.generated.schema.entity.services.connections.metadata.openMetadataConnection import (
+    OpenMetadataConnection,
+)
+from metadata.generated.schema.entity.services.connections.pipeline.databricksPipelineConnection import (
+    DatabricksPipelineConnection,
+)
+from metadata.generated.schema.metadataIngestion.workflow import (
+    Source as WorkflowSource,
+)
+from metadata.generated.schema.type.entityReference import EntityReference
+from metadata.ingestion.api.source import InvalidSourceException
+from metadata.ingestion.models.pipeline_status import OMetaPipelineStatus
+from metadata.ingestion.source.database.databricks.client import DatabricksClient
+from metadata.ingestion.source.pipeline.pipeline_service import PipelineServiceSource
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
+
+
+STATUS_MAP = {
+    "SUCCESS": StatusType.Successful,
+    "FAILED": StatusType.Failed,
+    "TIMEOUT": StatusType.Failed,
+    "CANCELED": StatusType.Failed,
+    "PENDING": StatusType.Pending,
+    "RUNNING": StatusType.Pending,
+    "TERMINATING": StatusType.Pending,
+    "SKIPPED": StatusType.Failed,
+    "INTERNAL_ERROR": StatusType.Failed,
+}
+
+TASK_MAP = {
+    "notebook_task": "Notebook_Task",
+    "spark_jar_task": "SparkJar_Task",
+    "spark_python_task": "SparkPython_Task",
+    "spark_submit_task": "SparkSubmit_Task",
+    "pipeline_task": "DeltaLiveTablePipeline_Task",
+    "python_wheel_task": "PythonWheel_Task",
+    "sql_task": "Sql_Task",
+    "dbt_task": "Dbt_Task",
+}
+
+
+class DatabrickspipelineSource(PipelineServiceSource):
+    """
+    Implements the necessary methods ot extract
+    Pipeline metadata from Databricks Jobs API
+    """
+
+    def __init__(self, config: WorkflowSource, metadata_config: OpenMetadataConnection):
+        super().__init__(config, metadata_config)
+        self.task_id_mapping = {}
+        self.job_id_list = set()
```

Comment:
Instead of setting the `self.job_id_list` here maybe we can use the `context` variable to handle the information. Might be a bit more explicit to help us know it is dependent on the execution state

Reply:
You can directly play with the context like this https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/src/metadata/ingestion/api/topology_runner.py#L142

Reply:
Good call! I updated the service to use the `context` variable

Reply:
thanks for updating!

The context is a pydantic BaseModel that allows arbitrary additions whenever it is used https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/src/metadata/ingestion/models/topology.py#L83

We use it in two ways:
1. Define the hierarchy of assets that are needed during the ingestion and keep a snapshot of "where we are" in the execution. E.g., to ingest a table, I first need to know its schema, database, and service. This information is embedded in the topology of the service itself in the `context` key https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/src/metadata/ingestion/source/database/database_service.py#L95
2. We pass parameters on the fly that are needed in different steps of the execution (example https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/src/metadata/ingestion/source/database/dbt/metadata.py#L290). To access/update the values you can either use it by name as your typical pydantic class `context.name` or you can jump straight into its `dict` structure if you need dynamism of some sorts `context.__dict__["name"]`. Here you could set `self.context.job_id_list = []` and then keep appending on the existing `self.context.job_id_list` during the loop. To pick it up in the status, just call `self.context.job_id_list`.

Hope this helps, this part is not really clear in the tutorials or docs as it's a rather internal detail so that's on us.

Feel free to keep asking until you have all the info you need.

Thanks!